{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0cca48fb",
   "metadata": {},
   "source": [
    "### Figure 5 master notebook\n",
    "\n",
    "What it does?\n",
    "1. Inside/Outside peak enrichment\n",
    "2. Model and ISM based prioritization\n",
    "3. Supplement: using random pli-matched genes\n",
    "\n",
    "Pre-requisite files\n",
    "- Gene list - From CGC: Only Oncogenes / TSGs (PIC: Adriana)\n",
    "- All genes and TSS - PromoterAI file (PIC: Nicole)\n",
    "- All vierstra arechetype ctcfs (PIC: Arvind)\n",
    "- Model scores (PIC: Arvind)\n",
    "- ISM scores (PIC: Neal)\n",
    "- Per cancer peakset (same as one used to train model) (PIC: Laksshman, copied over to Arvind's directory)\n",
    "- 5 fold models (PIC: Arvind)\n",
    "- Phylop file (PIC: Laksshman)\n",
    "- Fasta file (PIC: Laksshman)\n",
    "- PLI file: gnomad.v2.1.1.lof_metrics.by_gene.txt (PIC: Hong)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e31065a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "358467e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/akumar22/anaconda3/envs/python38/lib/python3.10/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from IPython.display import display, HTML\n",
    "import scipy.stats\n",
    "from pybedtools import BedTool\n",
    "import pickle\n",
    "from tqdm import tqdm\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "from seq2atac.stable import compute_gc_bed, one_hot_encode, read_pickle, compute_signal, write_pickle\n",
    "from seq2atac.analysis import get_promoterai_tss, get_cosmic_pancan_genes, fasta_file\n",
    "from seq2atac.analysis.enrichment_utils import create_pancancer_distribution_plots, create_pancancer_correlations, create_pancancer_valuecounts, create_pancancer_distribution_plots_discrete\n",
    "from seq2atac.analysis.sample_controls import matching_logic\n",
    "from seq2atac.analysis.mutation_utils import compute_vierstra_groups, compute_trinuc, ingene_indicator\n",
    "from seq2atac.analysis.mutation_processing_pipeline_utils import annotateNearestGene, annotateScore, filterannotatePeaks, annotatePhylopScore, annotatePhylopScorePeak, annotateBedfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0496d270",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctcf_vierstra_file = \"../process_tcga_icgc/ctcf_vierstra.bed\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d02e4c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "565"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pancan_genes = read_pickle(\"../process_tcga_icgc/ogtsg.pkl\")\n",
    "len(pancan_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1aee7355",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tcga_cancers = [\"BLCA\",\"BRCA\",\"GBM\",\"COAD\",\"KIRC\",\"KIRP\",\"LUAD\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be214b96",
   "metadata": {},
   "source": [
    "### Process all somatic mutations for LoF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "76d790ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "ism_scores = pd.read_csv(\"../score_backups/ism_scores_genomewide/all_cancers_all_mutations.csv\",index_col=0)\n",
    "\n",
    "def merge_ism_scores(somatic_df,ism_df):\n",
    "    \n",
    "    sh_before = len(somatic_df)\n",
    "    somatic_df = somatic_df.merge(ism_df,how=\"left\")\n",
    "    assert sh_before == somatic_df.shape[0]\n",
    "    assert len(somatic_df[somatic_df[\"ref\"].isna()]) == 0\n",
    "    return somatic_df\n",
    "\n",
    "# ism_scores[\"sample\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "72f00b5f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLCA\n",
      "(14219, 31)\n",
      "(14219, 58)\n",
      "(13593, 58)\n",
      "BRCA\n",
      "(38491, 31)\n",
      "(38491, 58)\n",
      "(36972, 58)\n",
      "GBM\n",
      "(5016, 31)\n",
      "(5016, 58)\n",
      "(4732, 58)\n",
      "COAD\n",
      "(40544, 31)\n",
      "(40544, 58)\n",
      "(39782, 58)\n",
      "KIRC\n",
      "(8361, 31)\n",
      "(8361, 58)\n",
      "(7984, 58)\n",
      "KIRP\n",
      "(3655, 31)\n",
      "(3655, 58)\n",
      "(3495, 58)\n",
      "LUAD\n",
      "(17010, 31)\n",
      "(17010, 58)\n",
      "(16028, 58)\n"
     ]
    }
   ],
   "source": [
    "somatic_df_dict = {}\n",
    "for cancer_name in all_tcga_cancers:\n",
    "    print(cancer_name)\n",
    "\n",
    "    ### read the processed somatic dataframes (this ensures within peaks)\n",
    "    somatic_df = read_pickle(f\"../somatic_filtered/{cancer_name}_filtered_annotated_somatic_regulatory.pkl\")\n",
    "    assert len(somatic_df.drop_duplicates([\"Chromosome\",\"hg38_start\",\"Tumor_Seq_Allele2\"])) == len(somatic_df)\n",
    "    print(somatic_df.shape)    \n",
    "    \n",
    "    \n",
    "    ### annotate model scores\n",
    "    \n",
    "    ### mutation centered scores\n",
    "    somatic_df = annotateScore(somatic_df,\n",
    "                               f\"../score_backups/summit_centered_peaks/{cancer_name}/fold_*_somatic_mutation_centered.pkl\")\n",
    "    somatic_df = somatic_df.rename(columns={\"proba_ref\":\"proba_ref_mutation_centered\",\n",
    "                                            \"proba_alt\":\"proba_alt_mutation_centered\"})\n",
    "    somatic_df[\"diff_mutation_centered\"] = somatic_df[\"proba_ref_mutation_centered\"] - somatic_df[\"proba_alt_mutation_centered\"]\n",
    "    \n",
    "    ### summit centered scores\n",
    "    somatic_df = annotateScore(somatic_df,\n",
    "                               f\"../score_backups/summit_centered_peaks/{cancer_name}/fold_*_somatic_summit_centered.pkl\",\n",
    "                              remove_foldwise=False)\n",
    "    somatic_df = somatic_df.rename(columns={\"proba_ref\":\"proba_ref_summit_centered\",\n",
    "                                            \"proba_alt\":\"proba_alt_summit_centered\"})\n",
    "    somatic_df[\"diff_summit_centered\"] = somatic_df[\"proba_ref_summit_centered\"] - somatic_df[\"proba_alt_summit_centered\"]\n",
    "    for fold in range(5):\n",
    "        somatic_df[f\"fold_{fold}_diff_summit_centered\"] = somatic_df[f\"proba_ref_{fold}\"] - somatic_df[f\"proba_alt_{fold}\"]\n",
    "\n",
    "    \n",
    "    ### annotate ism scores\n",
    "    somatic_df = merge_ism_scores(somatic_df,\n",
    "                                  ism_scores[ism_scores[\"sample\"]==cancer_name.lower()])\n",
    "    \n",
    "    \n",
    "    ### annotate pancan\n",
    "    somatic_df[\"closest_ogtsg\"] = somatic_df[\"gene\"].apply(lambda x : ingene_indicator(x, pancan_genes))\n",
    "\n",
    "    ### remove ctcfs\n",
    "    somatic_df = annotateBedfile(somatic_df,\n",
    "                                 ctcf_vierstra_file,\n",
    "                                 \"ctcf_uncleaned\")\n",
    "    print(somatic_df.shape)\n",
    "    somatic_df = somatic_df[somatic_df[\"ctcf_uncleaned\"]==0]\n",
    "    print(somatic_df.shape)\n",
    "    \n",
    "    \n",
    "    ### some annotations for matching\n",
    "    somatic_df[\"gc_2\"] = somatic_df[\"gc_peak\"].apply(lambda x:round(x,2))\n",
    "    somatic_df[\"gc_1\"] = somatic_df[\"gc_peak\"].apply(lambda x:round(x,1))\n",
    "    somatic_df[\"distance_to_summit_discrete\"] = somatic_df[\"distance_to_summit\"].apply(lambda x:round(x,-1))\n",
    "    \n",
    "    assert len(somatic_df.drop_duplicates([\"Chromosome\",\"hg38_start\",\"Tumor_Seq_Allele2\"])) == len(somatic_df)    \n",
    "    somatic_df_dict[cancer_name] = somatic_df\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8203bc0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BLCA': 0.047121298313140714,\n",
       " 'BRCA': 0.046678137779235754,\n",
       " 'GBM': 0.0649726867675781,\n",
       " 'COAD': 0.04962486773729317,\n",
       " 'KIRC': 0.046412371098995216,\n",
       " 'KIRP': 0.05527774393558505,\n",
       " 'LUAD': 0.0665462560951709}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Store these dataframes\n",
    "write_pickle(somatic_df_dict,\"somatic_df_dict_lof.pkl\")\n",
    "{cancer_name:somatic_df_dict[cancer_name][\"diff_summit_centered\"].quantile(0.975) for cancer_name in all_tcga_cancers}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d2c7338e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{0.01: {'BLCA': -0.057111942768096925, 'BRCA': -0.05661134839057922, 'GBM': -0.06666439831256867, 'COAD': -0.06200151786208153, 'KIRC': -0.048552260994911195, 'KIRP': -0.05301146030426025, 'LUAD': -0.07743662059307098}, 0.025: {'BLCA': -0.03862689733505249, 'BRCA': -0.036670146882534026, 'GBM': -0.04861255288124084, 'COAD': -0.04070858210325241, 'KIRC': -0.03324506804347038, 'KIRP': -0.03717219829559326, 'LUAD': -0.0480069562792778}, 0.05: {'BLCA': -0.027190136909484863, 'BRCA': -0.02477472126483917, 'GBM': -0.03535886406898499, 'COAD': -0.02590775489807129, 'KIRC': -0.021711236238479613, 'KIRP': -0.02503616511821747, 'LUAD': -0.029831188917160033}, 0.1: {'BLCA': -0.016697907447814943, 'BRCA': -0.01504697799682617, 'GBM': -0.0220944881439209, 'COAD': -0.01513502597808838, 'KIRC': -0.013126516342163086, 'KIRP': -0.013112854957580567, 'LUAD': -0.017016947269439697}, 0.9: {'BLCA': 0.016679382324218756, 'BRCA': 0.01682010293006897, 'GBM': 0.02704451084136965, 'COAD': 0.01715412735939026, 'KIRC': 0.01785607039928436, 'KIRP': 0.017917072772979735, 'LUAD': 0.02126397490501404}, 0.95: {'BLCA': 0.02902772426605218, 'BRCA': 0.029632744193077053, 'GBM': 0.04374605119228363, 'COAD': 0.030486193299293383, 'KIRC': 0.030386731028556793, 'KIRP': 0.03140127062797545, 'LUAD': 0.040654009580612174}, 0.975: {'BLCA': 0.047121298313140714, 'BRCA': 0.046678137779235754, 'GBM': 0.0649726867675781, 'COAD': 0.04962486773729317, 'KIRC': 0.046412371098995216, 'KIRP': 0.05527774393558505, 'LUAD': 0.0665462560951709}, 0.99: {'BLCA': 0.08331707358360288, 'BRCA': 0.07555846065282848, 'GBM': 0.10592751860618536, 'COAD': 0.08844304621219669, 'KIRC': 0.07949200928211213, 'KIRP': 0.09122762441635128, 'LUAD': 0.11240482211112943}}\n",
      "{0.01: {'BLCA': -0.12949332237243652, 'BRCA': -0.13525907099246978, 'GBM': -0.13547704100608826, 'COAD': -0.16648658692836762, 'KIRC': -0.12970117568969727, 'KIRP': -0.13647606670856477, 'LUAD': -0.1656234911084175}, 0.025: {'BLCA': -0.09360542893409729, 'BRCA': -0.09636679738759994, 'GBM': -0.09119054451584815, 'COAD': -0.11884650886058806, 'KIRC': -0.09568614140152931, 'KIRP': -0.10293575376272202, 'LUAD': -0.11393313482403755}, 0.05: {'BLCA': -0.06773808598518372, 'BRCA': -0.06810954362154006, 'GBM': -0.06794483661651611, 'COAD': -0.08838334083557128, 'KIRC': -0.06904576569795608, 'KIRP': -0.07765556275844573, 'LUAD': -0.08047639317810536}, 0.1: {'BLCA': -0.0450954794883728, 'BRCA': -0.04576551914215088, 'GBM': -0.04429733157157898, 'COAD': -0.060656607151031494, 'KIRC': -0.04769727289676666, 'KIRP': -0.054601073265075684, 'LUAD': -0.050702711939811705}, 0.9: {'BLCA': 0.050472068786621124, 'BRCA': 0.05158405900001526, 'GBM': 0.06165300309658064, 'COAD': 0.04980632364749908, 'KIRC': 0.05208898186683654, 'KIRP': 0.05353643000125885, 'LUAD': 0.06577606201171879}, 0.95: {'BLCA': 0.07354418039321899, 'BRCA': 0.07630454003810881, 'GBM': 0.09008108526468275, 'COAD': 0.07769675999879816, 'KIRC': 0.07946792989969254, 'KIRP': 0.0808287620544433, 'LUAD': 0.09811554625630377}, 0.975: {'BLCA': 0.10115094184875464, 'BRCA': 0.10343152731657014, 'GBM': 0.11602564603090282, 'COAD': 0.11152596399188042, 'KIRC': 0.10662183165550239, 'KIRP': 0.11235630959272387, 'LUAD': 0.13435505703091613}, 0.99: {'BLCA': 0.1576702582836151, 'BRCA': 0.14596161544322978, 'GBM': 0.16171254396438572, 'COAD': 0.16738623246550655, 'KIRC': 0.1565067245066166, 'KIRP': 0.1801906245946884, 'LUAD': 0.1890658167004585}}\n"
     ]
    }
   ],
   "source": [
    "thresholds = {}\n",
    "for q in [0.01,0.025,0.05,0.1,0.9,0.95,0.975,0.99]:\n",
    "    thresh_quant_dict = {cn:somatic_df_dict[cn][\"diff_summit_centered\"].quantile(q) for cn in all_tcga_cancers}\n",
    "    thresholds[q] = thresh_quant_dict\n",
    "print(thresholds)\n",
    "\n",
    "write_pickle(thresholds,\"threshold_summit_centered.pkl\")\n",
    "\n",
    "thresholds = {}\n",
    "for q in [0.01,0.025,0.05,0.1,0.9,0.95,0.975,0.99]:\n",
    "    thresh_quant_dict = {cn:somatic_df_dict[cn][\"diff_mutation_centered\"].quantile(q) for cn in all_tcga_cancers}\n",
    "    thresholds[q] = thresh_quant_dict\n",
    "print(thresholds)\n",
    "\n",
    "write_pickle(thresholds,\"threshold_mutation_centered.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d9e3d64",
   "metadata": {},
   "source": [
    "### Generate Matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "37f907d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_match(somatic_df_dict,genelist):\n",
    "    \n",
    "    all_tcga_cancers = list(somatic_df_dict.keys())\n",
    "    \n",
    "    ### gene annotation\n",
    "    for cancer_name in all_tcga_cancers:\n",
    "        somatic_df = somatic_df_dict[cancer_name].copy()\n",
    "        print(somatic_df.shape)\n",
    "\n",
    "        somatic_df[\"closest_pancan\"] = somatic_df[\"gene\"].apply(lambda x : ingene_indicator(x, genelist))\n",
    "\n",
    "        print(somatic_df.shape)\n",
    "        somatic_df_dict[cancer_name] = somatic_df   \n",
    "        \n",
    "    ### matching\n",
    "    pancan_matched_dict = {}\n",
    "    nonpancan_matched_dict = {}\n",
    "\n",
    "    for cancer_name in all_tcga_cancers:\n",
    "        print(cancer_name)\n",
    "        somatic_df = somatic_df_dict[cancer_name].copy()\n",
    "        df1, df2 = matching_logic(somatic_df[somatic_df[\"closest_pancan\"]==1],\n",
    "                                  somatic_df[(somatic_df[\"closest_pancan\"]==0) & (somatic_df[\"closest_ogtsg\"]==0)],\n",
    "                                  levels=[[\"trinuc\",\"gc_2\",\"distance_to_summit_discrete\"],\n",
    "                                         [\"trinuc\",\"gc_1\",\"distance_to_summit_discrete\"],\n",
    "                                         [\"trinuc\",\"distance_to_summit_discrete\"],\n",
    "                                         [\"distance_to_summit_discrete\"]],\n",
    "                                  verbose=False)\n",
    "\n",
    "        pancan_matched_dict[cancer_name] = df1\n",
    "        nonpancan_matched_dict[cancer_name] = df2\n",
    "        \n",
    "    ### sanity\n",
    "    df_stats = pd.DataFrame()\n",
    "    for cancer_name in all_tcga_cancers:\n",
    "        print(cancer_name)\n",
    "        somatic_df = somatic_df_dict[cancer_name].copy()\n",
    "        df_stats.loc[cancer_name,\"total somatic\"] = len(somatic_df)\n",
    "        df_stats.loc[cancer_name,\"pancan\"] = len(somatic_df[(somatic_df[\"closest_pancan\"]==1)])\n",
    "        df_stats.loc[cancer_name,\"nonpancan\"] = len(somatic_df[(somatic_df[\"closest_pancan\"]==0)])\n",
    "        df_stats.loc[cancer_name,\"matched pancan\"] = len(pancan_matched_dict[cancer_name])\n",
    "        df_stats.loc[cancer_name,\"matched not near pancan\"] = len(nonpancan_matched_dict[cancer_name])\n",
    "        \n",
    "        assert len(somatic_df[(somatic_df[\"closest_pancan\"]==1)]) == len(pancan_matched_dict[cancer_name])\n",
    "        assert len(pancan_matched_dict[cancer_name]) == len(nonpancan_matched_dict[cancer_name])\n",
    "        \n",
    "    display(df_stats.astype(int))\n",
    " \n",
    "    \n",
    "    return pancan_matched_dict, nonpancan_matched_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a74969cf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "565\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>888</td>\n",
       "      <td>12705</td>\n",
       "      <td>888</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2475</td>\n",
       "      <td>34497</td>\n",
       "      <td>2475</td>\n",
       "      <td>2475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>291</td>\n",
       "      <td>4441</td>\n",
       "      <td>291</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2351</td>\n",
       "      <td>37431</td>\n",
       "      <td>2351</td>\n",
       "      <td>2351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>484</td>\n",
       "      <td>7500</td>\n",
       "      <td>484</td>\n",
       "      <td>484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>230</td>\n",
       "      <td>3265</td>\n",
       "      <td>230</td>\n",
       "      <td>230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>952</td>\n",
       "      <td>15076</td>\n",
       "      <td>952</td>\n",
       "      <td>952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     888      12705             888   \n",
       "BRCA          36972    2475      34497            2475   \n",
       "GBM            4732     291       4441             291   \n",
       "COAD          39782    2351      37431            2351   \n",
       "KIRC           7984     484       7500             484   \n",
       "KIRP           3495     230       3265             230   \n",
       "LUAD          16028     952      15076             952   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      888  \n",
       "BRCA                     2475  \n",
       "GBM                       291  \n",
       "COAD                     2351  \n",
       "KIRC                      484  \n",
       "KIRP                      230  \n",
       "LUAD                      952  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(len(pancan_genes))\n",
    "pancan_matched_dict, nonpancan_matched_dict = get_match(somatic_df_dict,pancan_genes)\n",
    "write_pickle(pancan_matched_dict,\"pancan_matched_dict_ogtsg.pkl\")\n",
    "write_pickle(nonpancan_matched_dict,\"nonpancan_matched_dict_ogtsg.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bbfdabd",
   "metadata": {},
   "source": [
    "### 5 random pli matched genesets - also total mutations matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "430bd184",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>928</td>\n",
       "      <td>12665</td>\n",
       "      <td>928</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2432</td>\n",
       "      <td>34540</td>\n",
       "      <td>2432</td>\n",
       "      <td>2432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>332</td>\n",
       "      <td>4400</td>\n",
       "      <td>332</td>\n",
       "      <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2634</td>\n",
       "      <td>37148</td>\n",
       "      <td>2634</td>\n",
       "      <td>2634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>572</td>\n",
       "      <td>7412</td>\n",
       "      <td>572</td>\n",
       "      <td>572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>255</td>\n",
       "      <td>3240</td>\n",
       "      <td>255</td>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>1100</td>\n",
       "      <td>14928</td>\n",
       "      <td>1100</td>\n",
       "      <td>1100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     928      12665             928   \n",
       "BRCA          36972    2432      34540            2432   \n",
       "GBM            4732     332       4400             332   \n",
       "COAD          39782    2634      37148            2634   \n",
       "KIRC           7984     572       7412             572   \n",
       "KIRP           3495     255       3240             255   \n",
       "LUAD          16028    1100      14928            1100   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      928  \n",
       "BRCA                     2432  \n",
       "GBM                       332  \n",
       "COAD                     2634  \n",
       "KIRC                      572  \n",
       "KIRP                      255  \n",
       "LUAD                     1100  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>865</td>\n",
       "      <td>12728</td>\n",
       "      <td>865</td>\n",
       "      <td>865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2320</td>\n",
       "      <td>34652</td>\n",
       "      <td>2320</td>\n",
       "      <td>2320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>305</td>\n",
       "      <td>4427</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2534</td>\n",
       "      <td>37248</td>\n",
       "      <td>2534</td>\n",
       "      <td>2534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>564</td>\n",
       "      <td>7420</td>\n",
       "      <td>564</td>\n",
       "      <td>564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>234</td>\n",
       "      <td>3261</td>\n",
       "      <td>234</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>990</td>\n",
       "      <td>15038</td>\n",
       "      <td>990</td>\n",
       "      <td>990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     865      12728             865   \n",
       "BRCA          36972    2320      34652            2320   \n",
       "GBM            4732     305       4427             305   \n",
       "COAD          39782    2534      37248            2534   \n",
       "KIRC           7984     564       7420             564   \n",
       "KIRP           3495     234       3261             234   \n",
       "LUAD          16028     990      15038             990   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      865  \n",
       "BRCA                     2320  \n",
       "GBM                       305  \n",
       "COAD                     2534  \n",
       "KIRC                      564  \n",
       "KIRP                      234  \n",
       "LUAD                      990  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>964</td>\n",
       "      <td>12629</td>\n",
       "      <td>964</td>\n",
       "      <td>964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2558</td>\n",
       "      <td>34414</td>\n",
       "      <td>2558</td>\n",
       "      <td>2558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>380</td>\n",
       "      <td>4352</td>\n",
       "      <td>380</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2924</td>\n",
       "      <td>36858</td>\n",
       "      <td>2924</td>\n",
       "      <td>2924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>570</td>\n",
       "      <td>7414</td>\n",
       "      <td>570</td>\n",
       "      <td>570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>282</td>\n",
       "      <td>3213</td>\n",
       "      <td>282</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>1089</td>\n",
       "      <td>14939</td>\n",
       "      <td>1089</td>\n",
       "      <td>1089</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     964      12629             964   \n",
       "BRCA          36972    2558      34414            2558   \n",
       "GBM            4732     380       4352             380   \n",
       "COAD          39782    2924      36858            2924   \n",
       "KIRC           7984     570       7414             570   \n",
       "KIRP           3495     282       3213             282   \n",
       "LUAD          16028    1089      14939            1089   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      964  \n",
       "BRCA                     2558  \n",
       "GBM                       380  \n",
       "COAD                     2924  \n",
       "KIRC                      570  \n",
       "KIRP                      282  \n",
       "LUAD                     1089  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>874</td>\n",
       "      <td>12719</td>\n",
       "      <td>874</td>\n",
       "      <td>874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2562</td>\n",
       "      <td>34410</td>\n",
       "      <td>2562</td>\n",
       "      <td>2562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>375</td>\n",
       "      <td>4357</td>\n",
       "      <td>375</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2723</td>\n",
       "      <td>37059</td>\n",
       "      <td>2723</td>\n",
       "      <td>2723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>599</td>\n",
       "      <td>7385</td>\n",
       "      <td>599</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>273</td>\n",
       "      <td>3222</td>\n",
       "      <td>273</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>1142</td>\n",
       "      <td>14886</td>\n",
       "      <td>1142</td>\n",
       "      <td>1142</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     874      12719             874   \n",
       "BRCA          36972    2562      34410            2562   \n",
       "GBM            4732     375       4357             375   \n",
       "COAD          39782    2723      37059            2723   \n",
       "KIRC           7984     599       7385             599   \n",
       "KIRP           3495     273       3222             273   \n",
       "LUAD          16028    1142      14886            1142   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      874  \n",
       "BRCA                     2562  \n",
       "GBM                       375  \n",
       "COAD                     2723  \n",
       "KIRC                      599  \n",
       "KIRP                      273  \n",
       "LUAD                     1142  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>824</td>\n",
       "      <td>12769</td>\n",
       "      <td>824</td>\n",
       "      <td>824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2397</td>\n",
       "      <td>34575</td>\n",
       "      <td>2397</td>\n",
       "      <td>2397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>319</td>\n",
       "      <td>4413</td>\n",
       "      <td>319</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2636</td>\n",
       "      <td>37146</td>\n",
       "      <td>2636</td>\n",
       "      <td>2636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>518</td>\n",
       "      <td>7466</td>\n",
       "      <td>518</td>\n",
       "      <td>518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>236</td>\n",
       "      <td>3259</td>\n",
       "      <td>236</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>1080</td>\n",
       "      <td>14948</td>\n",
       "      <td>1080</td>\n",
       "      <td>1080</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     824      12769             824   \n",
       "BRCA          36972    2397      34575            2397   \n",
       "GBM            4732     319       4413             319   \n",
       "COAD          39782    2636      37146            2636   \n",
       "KIRC           7984     518       7466             518   \n",
       "KIRP           3495     236       3259             236   \n",
       "LUAD          16028    1080      14948            1080   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      824  \n",
       "BRCA                     2397  \n",
       "GBM                       319  \n",
       "COAD                     2636  \n",
       "KIRC                      518  \n",
       "KIRP                      236  \n",
       "LUAD                     1080  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    pli_matched_genes = read_pickle(f\"../process_tcga_icgc/pli_matched_genes_x2{i}.pkl\")\n",
    "    print(len(pli_matched_genes))\n",
    "    \n",
    "    assert len(set(pli_matched_genes).intersection(set(pancan_genes))) == 0\n",
    "    \n",
    "    pancan_matched_dict_pli, nonpancan_matched_dict_pli = get_match(somatic_df_dict,pli_matched_genes)\n",
    "    \n",
    "    write_pickle(pancan_matched_dict_pli,f\"pancan_matched_dict_pli_x2{i}.pkl\")\n",
    "    write_pickle(nonpancan_matched_dict_pli,f\"nonpancan_matched_dict_pli_x2{i}.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ae2e853a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in range(5):\n",
    "#     pli_matched_genes = read_pickle(f\"../process_tcga_icgc/pli_matched_tsgenes_x2{i}.pkl\")\n",
    "#     print(len(pli_matched_genes))\n",
    "    \n",
    "#     assert len(set(pli_matched_genes).intersection(set(pancan_genes))) == 0\n",
    "    \n",
    "#     pancan_matched_dict_pli, nonpancan_matched_dict_pli = get_match(somatic_df_dict,pli_matched_genes)\n",
    "    \n",
    "#     write_pickle(pancan_matched_dict_pli,f\"pancan_matched_dict_pli_ts_x2{i}.pkl\")\n",
    "#     write_pickle(nonpancan_matched_dict_pli,f\"nonpancan_matched_dict_pli_ts_x2{i}.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9778e3e1",
   "metadata": {},
   "source": [
    "### 5 random genesets - also total mutations matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "953b9e40",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>820</td>\n",
       "      <td>12773</td>\n",
       "      <td>820</td>\n",
       "      <td>820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2241</td>\n",
       "      <td>34731</td>\n",
       "      <td>2241</td>\n",
       "      <td>2241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>266</td>\n",
       "      <td>4466</td>\n",
       "      <td>266</td>\n",
       "      <td>266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2328</td>\n",
       "      <td>37454</td>\n",
       "      <td>2328</td>\n",
       "      <td>2328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>452</td>\n",
       "      <td>7532</td>\n",
       "      <td>452</td>\n",
       "      <td>452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>225</td>\n",
       "      <td>3270</td>\n",
       "      <td>225</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>966</td>\n",
       "      <td>15062</td>\n",
       "      <td>966</td>\n",
       "      <td>966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     820      12773             820   \n",
       "BRCA          36972    2241      34731            2241   \n",
       "GBM            4732     266       4466             266   \n",
       "COAD          39782    2328      37454            2328   \n",
       "KIRC           7984     452       7532             452   \n",
       "KIRP           3495     225       3270             225   \n",
       "LUAD          16028     966      15062             966   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      820  \n",
       "BRCA                     2241  \n",
       "GBM                       266  \n",
       "COAD                     2328  \n",
       "KIRC                      452  \n",
       "KIRP                      225  \n",
       "LUAD                      966  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>791</td>\n",
       "      <td>12802</td>\n",
       "      <td>791</td>\n",
       "      <td>791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2126</td>\n",
       "      <td>34846</td>\n",
       "      <td>2126</td>\n",
       "      <td>2126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>256</td>\n",
       "      <td>4476</td>\n",
       "      <td>256</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2408</td>\n",
       "      <td>37374</td>\n",
       "      <td>2408</td>\n",
       "      <td>2408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>524</td>\n",
       "      <td>7460</td>\n",
       "      <td>524</td>\n",
       "      <td>524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>181</td>\n",
       "      <td>3314</td>\n",
       "      <td>181</td>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>966</td>\n",
       "      <td>15062</td>\n",
       "      <td>966</td>\n",
       "      <td>966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     791      12802             791   \n",
       "BRCA          36972    2126      34846            2126   \n",
       "GBM            4732     256       4476             256   \n",
       "COAD          39782    2408      37374            2408   \n",
       "KIRC           7984     524       7460             524   \n",
       "KIRP           3495     181       3314             181   \n",
       "LUAD          16028     966      15062             966   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      791  \n",
       "BRCA                     2126  \n",
       "GBM                       256  \n",
       "COAD                     2408  \n",
       "KIRC                      524  \n",
       "KIRP                      181  \n",
       "LUAD                      966  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>770</td>\n",
       "      <td>12823</td>\n",
       "      <td>770</td>\n",
       "      <td>770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>1999</td>\n",
       "      <td>34973</td>\n",
       "      <td>1999</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>277</td>\n",
       "      <td>4455</td>\n",
       "      <td>277</td>\n",
       "      <td>277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2303</td>\n",
       "      <td>37479</td>\n",
       "      <td>2303</td>\n",
       "      <td>2303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>448</td>\n",
       "      <td>7536</td>\n",
       "      <td>448</td>\n",
       "      <td>448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>233</td>\n",
       "      <td>3262</td>\n",
       "      <td>233</td>\n",
       "      <td>233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>906</td>\n",
       "      <td>15122</td>\n",
       "      <td>906</td>\n",
       "      <td>906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     770      12823             770   \n",
       "BRCA          36972    1999      34973            1999   \n",
       "GBM            4732     277       4455             277   \n",
       "COAD          39782    2303      37479            2303   \n",
       "KIRC           7984     448       7536             448   \n",
       "KIRP           3495     233       3262             233   \n",
       "LUAD          16028     906      15122             906   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      770  \n",
       "BRCA                     1999  \n",
       "GBM                       277  \n",
       "COAD                     2303  \n",
       "KIRC                      448  \n",
       "KIRP                      233  \n",
       "LUAD                      906  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>773</td>\n",
       "      <td>12820</td>\n",
       "      <td>773</td>\n",
       "      <td>773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2271</td>\n",
       "      <td>34701</td>\n",
       "      <td>2271</td>\n",
       "      <td>2271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>292</td>\n",
       "      <td>4440</td>\n",
       "      <td>292</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2403</td>\n",
       "      <td>37379</td>\n",
       "      <td>2403</td>\n",
       "      <td>2403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>491</td>\n",
       "      <td>7493</td>\n",
       "      <td>491</td>\n",
       "      <td>491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>192</td>\n",
       "      <td>3303</td>\n",
       "      <td>192</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>1023</td>\n",
       "      <td>15005</td>\n",
       "      <td>1023</td>\n",
       "      <td>1023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     773      12820             773   \n",
       "BRCA          36972    2271      34701            2271   \n",
       "GBM            4732     292       4440             292   \n",
       "COAD          39782    2403      37379            2403   \n",
       "KIRC           7984     491       7493             491   \n",
       "KIRP           3495     192       3303             192   \n",
       "LUAD          16028    1023      15005            1023   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      773  \n",
       "BRCA                     2271  \n",
       "GBM                       292  \n",
       "COAD                     2403  \n",
       "KIRC                      491  \n",
       "KIRP                      192  \n",
       "LUAD                     1023  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(13593, 61)\n",
      "(13593, 61)\n",
      "(36972, 61)\n",
      "(36972, 61)\n",
      "(4732, 61)\n",
      "(4732, 61)\n",
      "(39782, 61)\n",
      "(39782, 61)\n",
      "(7984, 61)\n",
      "(7984, 61)\n",
      "(3495, 61)\n",
      "(3495, 61)\n",
      "(16028, 61)\n",
      "(16028, 61)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'gc_1', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['trinuc', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['distance_to_summit_discrete']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>13593</td>\n",
       "      <td>858</td>\n",
       "      <td>12735</td>\n",
       "      <td>858</td>\n",
       "      <td>858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>36972</td>\n",
       "      <td>2215</td>\n",
       "      <td>34757</td>\n",
       "      <td>2215</td>\n",
       "      <td>2215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4732</td>\n",
       "      <td>374</td>\n",
       "      <td>4358</td>\n",
       "      <td>374</td>\n",
       "      <td>374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>39782</td>\n",
       "      <td>2517</td>\n",
       "      <td>37265</td>\n",
       "      <td>2517</td>\n",
       "      <td>2517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>7984</td>\n",
       "      <td>469</td>\n",
       "      <td>7515</td>\n",
       "      <td>469</td>\n",
       "      <td>469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3495</td>\n",
       "      <td>200</td>\n",
       "      <td>3295</td>\n",
       "      <td>200</td>\n",
       "      <td>200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>16028</td>\n",
       "      <td>977</td>\n",
       "      <td>15051</td>\n",
       "      <td>977</td>\n",
       "      <td>977</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          13593     858      12735             858   \n",
       "BRCA          36972    2215      34757            2215   \n",
       "GBM            4732     374       4358             374   \n",
       "COAD          39782    2517      37265            2517   \n",
       "KIRC           7984     469       7515             469   \n",
       "KIRP           3495     200       3295             200   \n",
       "LUAD          16028     977      15051             977   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      858  \n",
       "BRCA                     2215  \n",
       "GBM                       374  \n",
       "COAD                     2517  \n",
       "KIRC                      469  \n",
       "KIRP                      200  \n",
       "LUAD                      977  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    random_genes = read_pickle(f\"../process_tcga_icgc/random_genes_{i}.pkl\")\n",
    "    print(len(random_genes))\n",
    "    \n",
    "    assert len(set(random_genes).intersection(set(pancan_genes))) == 0\n",
    "    \n",
    "    pancan_matched_dict_random, nonpancan_matched_dict_random = get_match(somatic_df_dict,random_genes)\n",
    "    \n",
    "    write_pickle(pancan_matched_dict_random,f\"pancan_matched_dict_random_{i}.pkl\")\n",
    "    write_pickle(nonpancan_matched_dict_random,f\"nonpancan_matched_dict_random_{i}.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8ff1e91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a8d4965",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
