{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library(BiocManager)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(ArchR)\n",
    "library(ggplot2)\n",
    "library(TFBSTools)\n",
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(harmony)\n",
    "library(Signac)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(JASPAR2018)\n",
    "library(edgeR)\n",
    "library(sctransform)\n",
    "data(\"geneAnnoHg38\")\n",
    "data(\"genomeAnnoHg38\")\n",
    "geneAnno <- geneAnnoHg38\n",
    "genomeAnno <- genomeAnnoHg38\n",
    "addArchRThreads(24)\n",
    "\n",
    "\n",
    "#may be dont run this for getting cnv calls\n",
    "fn <- unclass(lsf.str(envir = asNamespace(\"ArchR\"), all = TRUE))\n",
    "  for(i in seq_along(fn)){\n",
    "    tryCatch({\n",
    "      eval(parse(text=paste0(fn[i], '<-ArchR:::', fn[i])))\n",
    "    }, error = function(x){\n",
    "    })\n",
    "  }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Loading the archr project to compute the 10MB CNV-ATAC matrix \n",
    "proj <- loadArchRProject('ArchR project of interest')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#contains help functions computing the CNV matrix\n",
    "source(\"TCGA_Helper.R\")\n",
    "\n",
    "#adding a custom code in ArchR to add CNV matrix\n",
    ".addCNVMatrix<-function (i = NULL, ArrowFiles = NULL, normByNeighbors = FALSE, \n",
    "    cellNames = NULL, allCells = NULL, windows = NULL, force = FALSE, \n",
    "    tstart = NULL, subThreads = NULL, logFile = NULL) \n",
    "{\n",
    "    ArrowFile <- ArrowFiles[i]\n",
    "    o <- h5closeAll()\n",
    "    o <- h5closeAll()\n",
    "    o <- .createArrowGroup(ArrowFile = ArrowFile, group = \"Use_CNVMatrix_10MB_GCsmooth\", \n",
    "        force = force, logFile = logFile)\n",
    "    tstart <- Sys.time()\n",
    "    if (is.null(cellNames)) {\n",
    "        cellNames <- .availableCells(ArrowFile)\n",
    "    }\n",
    "    if (!is.null(allCells)) {\n",
    "        cellNames <- cellNames[cellNames %in% allCells]\n",
    "    }\n",
    "    uniqueChr <- as.character(unique(seqnames(windows)@values))\n",
    "    seWindows <- .safelapply(seq_along(uniqueChr), function(x) {\n",
    "        o <- h5closeAll()\n",
    "        chr <- uniqueChr[x]\n",
    "        windowsx <- windows[BiocGenerics::which(seqnames(windows) == \n",
    "            chr)]\n",
    "        rangesx <- ranges(windowsx)\n",
    "        .messageDiffTime(sprintf(\"Counting Windows for Chromosome %s of %s!\", \n",
    "            x, length(uniqueChr)), tstart)\n",
    "        fragments <- .getFragsFromArrow(ArrowFile, chr = chr, \n",
    "            out = \"IRanges\", cellNames = cellNames)\n",
    "        temp <- IRanges(start = start(fragments), width = 1)\n",
    "        stopifnot(length(temp) == length(fragments))\n",
    "        oleft <- findOverlaps(ranges(rangesx), temp)\n",
    "        oleft <- DataFrame(queryHits = Rle(queryHits(oleft)), \n",
    "            subjectHits = subjectHits(oleft))\n",
    "        temp <- IRanges(start = end(fragments), width = 1)\n",
    "        stopifnot(length(temp) == length(fragments))\n",
    "        oright <- findOverlaps(ranges(rangesx), temp)\n",
    "        oright <- DataFrame(queryHits = Rle(queryHits(oright)), \n",
    "            subjectHits = subjectHits(oright))\n",
    "        remove(temp)\n",
    "        oleft$subjectHits <- as.integer(BiocGenerics::match(mcols(fragments)$RG[oleft$subjectHits], \n",
    "            cellNames))\n",
    "        oright$subjectHits <- as.integer(BiocGenerics::match(mcols(fragments)$RG[oright$subjectHits], \n",
    "            cellNames))\n",
    "        remove(fragments)\n",
    "        #\n",
    "        mat <- Matrix::sparseMatrix(i = c(oleft$queryHits, oright$queryHits), \n",
    "            j = c(oleft$subjectHits, oright$subjectHits), x = rep(1, \n",
    "                nrow(oleft) + nrow(oright)), dims = c(length(rangesx), \n",
    "                length(cellNames)))\n",
    "        colnames(mat) <- cellNames\n",
    "        windowSummary <- GRanges()\n",
    "        countSummary <- matrix(nrow = length(unique(mcols(windowsx)$name)), \n",
    "            ncol = ncol(mat))\n",
    "        rownames(countSummary) <- unique(mcols(windowsx)$name)\n",
    "        colnames(countSummary) <- cellNames\n",
    "        for (y in seq_len(nrow(countSummary))) {\n",
    "            idx <- which(mcols(windowsx)$name == rownames(countSummary)[y])\n",
    "            wx <- windowsx[idx]\n",
    "            wo <- GRanges(mcols(wx)$wSeq, ranges = IRanges(mcols(wx)$wStart, \n",
    "                mcols(wx)$wEnd))[1, ]\n",
    "            mcols(wo)$name <- mcols(wx)$name[1]\n",
    "            mcols(wo)$effectiveLength <- sum(width(wx))\n",
    "            mcols(wo)$percentEffectiveLength <- 100 * sum(width(wx))/width(wo)\n",
    "            mcols(wo)$GC <- sum(mcols(wx)$GC * width(wx))/width(wo)\n",
    "            mcols(wo)$AT <- sum(mcols(wx)$AT * width(wx))/width(wo)\n",
    "            mcols(wo)$N <- sum(mcols(wx)$N * width(wx))/width(wo)\n",
    "            countSummary[y, ] <- Matrix::colSums(mat[idx, , drop = FALSE])/sum(width(wx))\n",
    "            windowSummary <- c(windowSummary, wo)\n",
    "        }\n",
    "        seqlevels(windowSummary) <- uniqueChr\n",
    "        SummarizedExperiment::SummarizedExperiment(assays = SimpleList(counts = countSummary), \n",
    "            rowData = windowSummary)\n",
    "    }, threads = 1) %>% Reduce(\"rbind\", .)\n",
    "    .messageDiffTime(\"Filtering Low Quality Windows\", tstart)\n",
    "    seWindows <- seWindows[which(mcols(seWindows)$N < 0.001)]\n",
    "    seWindows <- seWindows[rowData(seWindows)$percentEffectiveLength >= \n",
    "        90]\n",
    "    .messageDiffTime(\"Normalizing GC-Bias\", tstart)\n",
    "    k <- 25\n",
    "    seWindows <- seWindows[order(mcols(seWindows)$GC)]\n",
    "    assays(seWindows)$log2GCNorm <- log2((assays(seWindows)$counts + \n",
    "        1e-05)/(apply(assays(seWindows)$counts, 2, function(x) .centerRollMean(x, \n",
    "        k)) + 1e-05))\n",
    "    if (normByNeighbors) {\n",
    "        .messageDiffTime(\"Normalizing Coverage-Bias using Neighbors\", \n",
    "            tstart)\n",
    "        totalCounts <- colSums(assays(seWindows)$counts * rowData(seWindows)$effectiveLength)\n",
    "        seWindows <- seWindows[, order(totalCounts)]\n",
    "        assays(seWindows)$log2GCNorm <- assays(seWindows)$log2GCNorm - \n",
    "            t(apply(assays(seWindows)$log2GCNorm, 1, function(y) .centerRollMean(y, \n",
    "                floor(0.025 * ncol(seWindows)))))\n",
    "        seWindows <- seWindows[, cellNames]\n",
    "    }\n",
    "    .messageDiffTime(\"Smoothing CNV Scores\", tstart)\n",
    "    seWindows <- sort(sortSeqlevels(seWindows))\n",
    "    temp_smooth<-apply(assays(seWindows)$log2GCNorm, \n",
    "            2, function(y) .centerRollMean(y, ceiling((windowSize/stepSize)/2)))\n",
    "    assays(seWindows,withDimnames=FALSE)$log2GCSmooth<-temp_smooth\n",
    "    windows <- lapply(split(rowRanges(seWindows), seqnames(seWindows)), \n",
    "        function(x) {\n",
    "            mcols(x)$idx <- seq_along(x)\n",
    "            x\n",
    "        }) %>% Reduce(\"c\", .) %>% sortSeqlevels %>% sort\n",
    "    rowRanges(seWindows) <- windows[rownames(seWindows)]\n",
    "    .messageDiffTime(\"Adding to Arrow Files\", tstart)\n",
    "    dfParams <- data.frame(windowSize = windowSize, stepSize = stepSize, excludeChr = c(\"chrM\",\"chrY\"),\n",
    "         stringsAsFactors = FALSE)\n",
    "    featureDF <- data.frame(seqnames = paste0(seqnames(seWindows)), \n",
    "        idx = mcols(seWindows)$name, start = start(seWindows), \n",
    "        end = end(seWindows), name = mcols(seWindows)$name, GC = mcols(seWindows)$GC, \n",
    "        effectiveLength = mcols(seWindows)$effectiveLength, stringsAsFactors = FALSE)\n",
    "    featureDF$idx<-seq_len(nrow(featureDF))\n",
    "    featureDF1<-featureDF          \n",
    "    featureDF <- do.call(\n",
    "      rbind.data.frame,\n",
    "      by(featureDF1, featureDF1$seqnames, function(df) { df$idx <- seq_len(nrow(df)); df; })\n",
    "    )\n",
    "    rownames(featureDF)<-featureDF$name\n",
    "    o <- .initializeMat(ArrowFile = ArrowFile, Group = \"Use_CNVMatrix_10MB_GCsmooth\", \n",
    "        Class = \"Double\", cellNames = cellNames, params = dfParams, \n",
    "        featureDF = featureDF, force = force)\n",
    "    uniqueChr <- unique(paste0(seqnames(seWindows)))\n",
    "    seWindows <- seWindows[, cellNames]\n",
    "    rownames(seWindows)<-rownames(featureDF)\n",
    "    rowData(seWindows)<-featureDF\n",
    "    for (x in seq_along(uniqueChr)) {\n",
    "        print(x)\n",
    "         o <- .addMatToArrow(mat = Matrix::Matrix(assays(seWindows[BiocGenerics::which(seqnames(seWindows) == \n",
    "            uniqueChr[x])])$log2GCSmooth, sparse = TRUE), ArrowFile = ArrowFile, \n",
    "            Group = paste0(\"Use_CNVMatrix_10MB_GCsmooth/\", uniqueChr[x]), binarize = FALSE)\n",
    "        gc()\n",
    "    }\n",
    "    return(ArrowFile)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "windowSize = 10e+06\n",
    "stepSize = 20e+05\n",
    "\n",
    "out <- addCNVMatrix(\n",
    "\tinput = getArrowFiles(proj),\n",
    "\tchromSizes = getChromSizes(proj), \n",
    "\tblacklist = getBlacklist(proj), \n",
    "\tgenome = getGenome(proj),\n",
    "\tforce = TRUE,windowSize = 10e+06, stepSize = 20e+05,\n",
    "    threads=20\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Plotting some regions of the 10MB CNV atac matrix\n",
    "proj <- addImputeWeights(ArchRProj = proj)\n",
    "\n",
    "markerGenes  <- c(\"w21\" )\n",
    "\n",
    "\n",
    "#Plot the UMAP Embedding with Marker Genes Overlayed w/ Imputation\n",
    "plotList<-plotEmbedding(ArchRProj = proj, colorBy = \"Use_CNVMatrix_10MB_GCsmooth\",pal = paletteContinuous(set = \"coolwarm\"), name = markerGenes, imputeWeights = getImputeWeights(proj))\n",
    "plotList\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveArchRProject(proj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
