{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1afb8511",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "from pyfaidx import Fasta\n",
    "from seq2atac.analysis import fasta_file\n",
    "fasta_seq=Fasta(fasta_file)\n",
    "\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19adc28d",
   "metadata": {},
   "outputs": [],
   "source": [
    "usecols = [\"Chromosome\",\"Start_position\",\"End_position\",\"Strand\",\"Variant_Classification\",\"Variant_Type\",\n",
    "           \"Reference_Allele\",\"Tumor_Seq_Allele2\",\"i_NumCallers\",\"i_repeat_masker\",\"Project_Code\",\"Donor_ID\"\n",
    "           ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8e5da3f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_files = {\"ICGC\": \"/illumina/scratch/deep_learning/public_data/ICGC/release_28/Consensus-SNV/final_consensus_passonly.snv_mnv_indel.icgc.public.maf.gz\",\n",
    "                  \"TCGA\": \"/illumina/scratch/deep_learning/public_data/ICGC/release_28/Consensus-SNV/final_consensus_passonly.snv_mnv_indel.tcga.controlled.maf.gz\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "abfe8121",
   "metadata": {},
   "outputs": [],
   "source": [
    "histology_df = pd.read_csv(\"/illumina/scratch/deep_learning/public_data/ICGC/release_28/summary/pcawg_specimen_histology_August2016_v9.csv\")\n",
    "donorid_to_project_dict = dict(zip(histology_df[\"icgc_donor_id\"].tolist(), histology_df[\"project_code\"].tolist()))\n",
    "donorid_to_histology_dict = dict(zip(histology_df[\"icgc_donor_id\"].tolist(), histology_df[\"histology_abbreviation\"].tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f56b66f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>i_NumCallers</th>\n",
       "      <th>i_repeat_masker</th>\n",
       "      <th>Project_Code</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>project_code</th>\n",
       "      <th>histology_abbreviation</th>\n",
       "      <th>mutation_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1230448</td>\n",
       "      <td>1230448</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1609723</td>\n",
       "      <td>1609723</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1903276</td>\n",
       "      <td>1903276</td>\n",
       "      <td>+</td>\n",
       "      <td>IGR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2574999</td>\n",
       "      <td>2574999</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3151250</td>\n",
       "      <td>3151250</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470627</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154769962</td>\n",
       "      <td>154769962</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>L1M4c</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470628</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154771230</td>\n",
       "      <td>154771230</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>L1PA7</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470629</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154771597</td>\n",
       "      <td>154771597</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>L1PA7</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470630</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154772343</td>\n",
       "      <td>154772343</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>L1MCb</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470631</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155231514</td>\n",
       "      <td>155231514</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41470632 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Start_position  End_position Strand  \\\n",
       "0              chr1         1230448       1230448      +   \n",
       "1              chr1         1609723       1609723      +   \n",
       "2              chr1         1903276       1903276      +   \n",
       "3              chr1         2574999       2574999      +   \n",
       "4              chr1         3151250       3151250      +   \n",
       "...             ...             ...           ...    ...   \n",
       "41470627       chrX       154769962     154769962      +   \n",
       "41470628       chrX       154771230     154771230      +   \n",
       "41470629       chrX       154771597     154771597      +   \n",
       "41470630       chrX       154772343     154772343      +   \n",
       "41470631       chrX       155231514     155231514      +   \n",
       "\n",
       "         Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0                        Intron          SNP                G   \n",
       "1                        Intron          SNP                C   \n",
       "2                           IGR          SNP                C   \n",
       "3                        Intron          SNP                C   \n",
       "4                        Intron          SNP                G   \n",
       "...                         ...          ...              ...   \n",
       "41470627                 Intron          SNP                G   \n",
       "41470628                 Intron          SNP                G   \n",
       "41470629                 Intron          SNP                G   \n",
       "41470630                 Intron          SNP                G   \n",
       "41470631                 Intron          SNP                C   \n",
       "\n",
       "         Tumor_Seq_Allele2 i_NumCallers i_repeat_masker   Project_Code  \\\n",
       "0                        A            4             NaN  Ovary-AdenoCA   \n",
       "1                        T            4             NaN  Ovary-AdenoCA   \n",
       "2                        T            4             NaN  Ovary-AdenoCA   \n",
       "3                        T            4             NaN  Ovary-AdenoCA   \n",
       "4                        A            4             NaN  Ovary-AdenoCA   \n",
       "...                    ...          ...             ...            ...   \n",
       "41470627                 A            4           L1M4c    Bladder-TCC   \n",
       "41470628                 C            4           L1PA7    Bladder-TCC   \n",
       "41470629                 T            3           L1PA7    Bladder-TCC   \n",
       "41470630                 T            4           L1MCb    Bladder-TCC   \n",
       "41470631                 T            4             NaN    Bladder-TCC   \n",
       "\n",
       "         Donor_ID project_code histology_abbreviation mutation_set  \n",
       "0         DO46416        OV-AU          Ovary-AdenoCA         ICGC  \n",
       "1         DO46416        OV-AU          Ovary-AdenoCA         ICGC  \n",
       "2         DO46416        OV-AU          Ovary-AdenoCA         ICGC  \n",
       "3         DO46416        OV-AU          Ovary-AdenoCA         ICGC  \n",
       "4         DO46416        OV-AU          Ovary-AdenoCA         ICGC  \n",
       "...           ...          ...                    ...          ...  \n",
       "41470627    DO720      BLCA-US            Bladder-TCC         TCGA  \n",
       "41470628    DO720      BLCA-US            Bladder-TCC         TCGA  \n",
       "41470629    DO720      BLCA-US            Bladder-TCC         TCGA  \n",
       "41470630    DO720      BLCA-US            Bladder-TCC         TCGA  \n",
       "41470631    DO720      BLCA-US            Bladder-TCC         TCGA  \n",
       "\n",
       "[41470632 rows x 15 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_rows = []\n",
    "for mutation_set,all_mutations in mutation_files.items():\n",
    "    for chunk in pd.read_csv(all_mutations, sep=\"\\t\", chunksize=1e6, usecols=usecols,low_memory=False):\n",
    "        chunk = chunk[(chunk[\"Variant_Type\"]==\"SNP\") & \n",
    "                      (chunk[\"Variant_Classification\"].isin([\"IGR\",\"Intron\"]))]\n",
    "\n",
    "        ### modify chromosome to be an actual number\n",
    "        chunk[\"Chromosome\"] = chunk[\"Chromosome\"].apply(lambda num:f\"chr{num}\")\n",
    "        ### based on donor to cancer dict, add type here\n",
    "        chunk[\"project_code\"] = chunk[\"Donor_ID\"].apply(lambda donor_id:donorid_to_project_dict.get(donor_id,\"None\"))\n",
    "        chunk[\"histology_abbreviation\"] = chunk[\"Donor_ID\"].apply(lambda donor_id:donorid_to_histology_dict.get(donor_id,\"None\"))\n",
    "        chunk[\"mutation_set\"] = mutation_set\n",
    "        all_rows.append(chunk)\n",
    "\n",
    "        print(len(all_rows))\n",
    "   \n",
    "\n",
    "     \n",
    "mutations_df = pd.concat(all_rows,axis=0,ignore_index=True)\n",
    "mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "62953164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2658,)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_df[\"Donor_ID\"].unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4812bc0",
   "metadata": {},
   "source": [
    "### Visualize some stats and add cancer code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9afae29e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mutations_df.groupby([\"project_code\",\"histology_abbreviation\"]).size().reset_index(name=\"counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e4bf6235",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>i_NumCallers</th>\n",
       "      <th>i_repeat_masker</th>\n",
       "      <th>Project_Code</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>project_code</th>\n",
       "      <th>histology_abbreviation</th>\n",
       "      <th>mutation_set</th>\n",
       "      <th>cancer_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1230448</td>\n",
       "      <td>1230448</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1609723</td>\n",
       "      <td>1609723</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1903276</td>\n",
       "      <td>1903276</td>\n",
       "      <td>+</td>\n",
       "      <td>IGR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2574999</td>\n",
       "      <td>2574999</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3151250</td>\n",
       "      <td>3151250</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>DO46416</td>\n",
       "      <td>OV-AU</td>\n",
       "      <td>Ovary-AdenoCA</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470627</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154769962</td>\n",
       "      <td>154769962</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>L1M4c</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470628</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154771230</td>\n",
       "      <td>154771230</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>L1PA7</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470629</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154771597</td>\n",
       "      <td>154771597</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>L1PA7</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470630</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154772343</td>\n",
       "      <td>154772343</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>L1MCb</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41470631</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155231514</td>\n",
       "      <td>155231514</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41470632 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Start_position  End_position Strand  \\\n",
       "0              chr1         1230448       1230448      +   \n",
       "1              chr1         1609723       1609723      +   \n",
       "2              chr1         1903276       1903276      +   \n",
       "3              chr1         2574999       2574999      +   \n",
       "4              chr1         3151250       3151250      +   \n",
       "...             ...             ...           ...    ...   \n",
       "41470627       chrX       154769962     154769962      +   \n",
       "41470628       chrX       154771230     154771230      +   \n",
       "41470629       chrX       154771597     154771597      +   \n",
       "41470630       chrX       154772343     154772343      +   \n",
       "41470631       chrX       155231514     155231514      +   \n",
       "\n",
       "         Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0                        Intron          SNP                G   \n",
       "1                        Intron          SNP                C   \n",
       "2                           IGR          SNP                C   \n",
       "3                        Intron          SNP                C   \n",
       "4                        Intron          SNP                G   \n",
       "...                         ...          ...              ...   \n",
       "41470627                 Intron          SNP                G   \n",
       "41470628                 Intron          SNP                G   \n",
       "41470629                 Intron          SNP                G   \n",
       "41470630                 Intron          SNP                G   \n",
       "41470631                 Intron          SNP                C   \n",
       "\n",
       "         Tumor_Seq_Allele2 i_NumCallers i_repeat_masker   Project_Code  \\\n",
       "0                        A            4             NaN  Ovary-AdenoCA   \n",
       "1                        T            4             NaN  Ovary-AdenoCA   \n",
       "2                        T            4             NaN  Ovary-AdenoCA   \n",
       "3                        T            4             NaN  Ovary-AdenoCA   \n",
       "4                        A            4             NaN  Ovary-AdenoCA   \n",
       "...                    ...          ...             ...            ...   \n",
       "41470627                 A            4           L1M4c    Bladder-TCC   \n",
       "41470628                 C            4           L1PA7    Bladder-TCC   \n",
       "41470629                 T            3           L1PA7    Bladder-TCC   \n",
       "41470630                 T            4           L1MCb    Bladder-TCC   \n",
       "41470631                 T            4             NaN    Bladder-TCC   \n",
       "\n",
       "         Donor_ID project_code histology_abbreviation mutation_set cancer_code  \n",
       "0         DO46416        OV-AU          Ovary-AdenoCA         ICGC          OV  \n",
       "1         DO46416        OV-AU          Ovary-AdenoCA         ICGC          OV  \n",
       "2         DO46416        OV-AU          Ovary-AdenoCA         ICGC          OV  \n",
       "3         DO46416        OV-AU          Ovary-AdenoCA         ICGC          OV  \n",
       "4         DO46416        OV-AU          Ovary-AdenoCA         ICGC          OV  \n",
       "...           ...          ...                    ...          ...         ...  \n",
       "41470627    DO720      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "41470628    DO720      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "41470629    DO720      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "41470630    DO720      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "41470631    DO720      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "\n",
       "[41470632 rows x 16 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_df[\"cancer_code\"] = mutations_df[\"project_code\"].apply(lambda string:string.split(\"-\")[0].upper())\n",
    "mutations_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c82f46be",
   "metadata": {},
   "source": [
    "### Donor Selection\n",
    "\n",
    "1. Keep only [\"BLCA\",\"BRCA\",\"GBM\",\"COAD\",\"KIRC\",\"KIRP\",\"LUAD\",\"SKCM\",\"RECA\",\"MELA\"] cancer codes\n",
    "2. Within BRCA, remove histologies [\"Breast-DCIS\",\"Breast-LobularCA\"]\n",
    "3. For COAD, remove MSI samples. please pay attention to the outputs of the cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d84a2915",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20348478, 16)\n"
     ]
    }
   ],
   "source": [
    "valid_cancer_codes = [\"BLCA\",\"BRCA\",\"GBM\",\"COAD\",\"KIRC\",\"KIRP\",\"LUAD\",\"SKCM\",\"RECA\",\"MELA\"]\n",
    "invalid_histologies = [\"Breast-DCIS\",\"Breast-LobularCA\"]\n",
    "mutations_df = mutations_df[mutations_df[\"cancer_code\"].isin(valid_cancer_codes)]\n",
    "mutations_df = mutations_df[~mutations_df[\"histology_abbreviation\"].isin(invalid_histologies)]\n",
    "print(mutations_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "24f9d1e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['COAD']\n",
      "After removing MSI samples (COAD):  (19682242, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>i_NumCallers</th>\n",
       "      <th>i_repeat_masker</th>\n",
       "      <th>Project_Code</th>\n",
       "      <th>Donor_ID</th>\n",
       "      <th>project_code</th>\n",
       "      <th>histology_abbreviation</th>\n",
       "      <th>mutation_set</th>\n",
       "      <th>cancer_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1489107</td>\n",
       "      <td>1489107</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>AluY</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>DO46853</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>RECA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2087940</td>\n",
       "      <td>2087940</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>DO46853</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>RECA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4329055</td>\n",
       "      <td>4329055</td>\n",
       "      <td>+</td>\n",
       "      <td>IGR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>DO46853</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>RECA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5165802</td>\n",
       "      <td>5165802</td>\n",
       "      <td>+</td>\n",
       "      <td>IGR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>MLT1I</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>DO46853</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>RECA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5822758</td>\n",
       "      <td>5822758</td>\n",
       "      <td>+</td>\n",
       "      <td>IGR</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>3</td>\n",
       "      <td>AluJr4</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>DO46853</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>Kidney-RCC</td>\n",
       "      <td>ICGC</td>\n",
       "      <td>RECA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19682237</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154769962</td>\n",
       "      <td>154769962</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>L1M4c</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19682238</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154771230</td>\n",
       "      <td>154771230</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>L1PA7</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19682239</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154771597</td>\n",
       "      <td>154771597</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>3</td>\n",
       "      <td>L1PA7</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19682240</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154772343</td>\n",
       "      <td>154772343</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>L1MCb</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19682241</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155231514</td>\n",
       "      <td>155231514</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>DO720</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>Bladder-TCC</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>BLCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19682242 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Chromosome  Start_position  End_position Strand  \\\n",
       "0              chr1         1489107       1489107      +   \n",
       "1              chr1         2087940       2087940      +   \n",
       "2              chr1         4329055       4329055      +   \n",
       "3              chr1         5165802       5165802      +   \n",
       "4              chr1         5822758       5822758      +   \n",
       "...             ...             ...           ...    ...   \n",
       "19682237       chrX       154769962     154769962      +   \n",
       "19682238       chrX       154771230     154771230      +   \n",
       "19682239       chrX       154771597     154771597      +   \n",
       "19682240       chrX       154772343     154772343      +   \n",
       "19682241       chrX       155231514     155231514      +   \n",
       "\n",
       "         Variant_Classification Variant_Type Reference_Allele  \\\n",
       "0                        Intron          SNP                C   \n",
       "1                        Intron          SNP                G   \n",
       "2                           IGR          SNP                C   \n",
       "3                           IGR          SNP                C   \n",
       "4                           IGR          SNP                T   \n",
       "...                         ...          ...              ...   \n",
       "19682237                 Intron          SNP                G   \n",
       "19682238                 Intron          SNP                G   \n",
       "19682239                 Intron          SNP                G   \n",
       "19682240                 Intron          SNP                G   \n",
       "19682241                 Intron          SNP                C   \n",
       "\n",
       "         Tumor_Seq_Allele2 i_NumCallers i_repeat_masker Project_Code Donor_ID  \\\n",
       "0                        T            4            AluY   Kidney-RCC  DO46853   \n",
       "1                        T            4             NaN   Kidney-RCC  DO46853   \n",
       "2                        T            4             NaN   Kidney-RCC  DO46853   \n",
       "3                        A            4           MLT1I   Kidney-RCC  DO46853   \n",
       "4                        G            3          AluJr4   Kidney-RCC  DO46853   \n",
       "...                    ...          ...             ...          ...      ...   \n",
       "19682237                 A            4           L1M4c  Bladder-TCC    DO720   \n",
       "19682238                 C            4           L1PA7  Bladder-TCC    DO720   \n",
       "19682239                 T            3           L1PA7  Bladder-TCC    DO720   \n",
       "19682240                 T            4           L1MCb  Bladder-TCC    DO720   \n",
       "19682241                 T            4             NaN  Bladder-TCC    DO720   \n",
       "\n",
       "         project_code histology_abbreviation mutation_set cancer_code  \n",
       "0             RECA-EU             Kidney-RCC         ICGC        RECA  \n",
       "1             RECA-EU             Kidney-RCC         ICGC        RECA  \n",
       "2             RECA-EU             Kidney-RCC         ICGC        RECA  \n",
       "3             RECA-EU             Kidney-RCC         ICGC        RECA  \n",
       "4             RECA-EU             Kidney-RCC         ICGC        RECA  \n",
       "...               ...                    ...          ...         ...  \n",
       "19682237      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "19682238      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "19682239      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "19682240      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "19682241      BLCA-US            Bladder-TCC         TCGA        BLCA  \n",
       "\n",
       "[19682242 rows x 16 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coad_remove_df = pd.read_csv(\"/illumina/scratch/deep_learning/akumar22/TCGA/mutations_scoring/master_files/msi_pcawg_samples.csv\",sep=\"\\t\")\n",
    "ids_to_remove = coad_remove_df[\"icgc_donor_id\"].tolist()\n",
    "print(mutations_df[mutations_df[\"Donor_ID\"].isin(ids_to_remove)][\"cancer_code\"].unique())\n",
    "\n",
    "### make sure its only coad\n",
    "mutations_df = mutations_df[~mutations_df[\"Donor_ID\"].isin(ids_to_remove)]\n",
    "print(\"After removing MSI samples (COAD): \",mutations_df.shape)\n",
    "\n",
    "mutations_df = mutations_df.reset_index(drop=True)\n",
    "mutations_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "50b4559e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MELA    6682332\n",
       "COAD    5247285\n",
       "SKCM    3736996\n",
       "BRCA    1208987\n",
       "LUAD    1154611\n",
       "RECA     457284\n",
       "GBM      448066\n",
       "BLCA     430742\n",
       "KIRC     168147\n",
       "KIRP     147792\n",
       "Name: cancer_code, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_df[\"cancer_code\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8780fcb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(588,)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_df[\"Donor_ID\"].unique().shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a9d9519",
   "metadata": {},
   "source": [
    "### Convert ref, alt to upper case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a083999",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_df[\"Reference_Allele\"] = mutations_df[\"Reference_Allele\"].apply(lambda x : x.upper())\n",
    "mutations_df[\"Tumor_Seq_Allele2\"] = mutations_df[\"Tumor_Seq_Allele2\"].apply(lambda x : x.upper())\n",
    "mutations_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab54e346",
   "metadata": {},
   "source": [
    "### hg19 to hg38 function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e31cf4ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from seq2atac.analysis.mutation_processing_pipeline_utils import hg19_to_hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da25a9dc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mutations_df_hg38 = hg19_to_hg38(mutations_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b71a53",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_df_hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5d4b6fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_df_hg38.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d9cb2cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_cols = [\"Chromosome\",\"hg38_start\",\"hg38_end\",\"Reference_Allele\",\"Tumor_Seq_Allele2\"]\n",
    "other_cols = [col for col in mutations_df_hg38.columns if col not in bed_cols]\n",
    "bed_cols, other_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d907a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_df_hg38 = mutations_df_hg38[bed_cols + other_cols]\n",
    "mutations_df_hg38"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64304d57",
   "metadata": {},
   "source": [
    "### Create cancer specific files along with de-duplication annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d201b34",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tcga_cancers = [\"BLCA\",\"BRCA\",\"GBM\",\"COAD\",\"KIRC\",\"KIRP\",\"LUAD\",\"SKCM\"]\n",
    "cancer_to_codes = {\"BLCA\":[\"BLCA\"],\n",
    "                   \"BRCA\":[\"BRCA\"],\n",
    "                   \"GBM\":[\"GBM\"],\n",
    "                   \"COAD\":[\"COAD\"], ## this could be coad, read.\n",
    "                   \"KIRC\":[\"KIRC\",\"RECA\"],\n",
    "                   \"KIRP\":[\"KIRP\"],\n",
    "                   \"LUAD\":[\"LUAD\"],\n",
    "                   \"SKCM\":[\"SKCM\",\"MELA\"]\n",
    "                   }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "757fd910",
   "metadata": {},
   "outputs": [],
   "source": [
    "from seq2atac.stable import write_pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfc8c7ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!mkdir cancer_specific_somatic_hg38/\n",
    "for cancer_name in all_tcga_cancers:\n",
    "    print(cancer_name)\n",
    "\n",
    "    # which codes can be scored by a particular TCGA model\n",
    "    cancer_types = cancer_to_codes[cancer_name]\n",
    "    print(cancer_types)\n",
    "    cancer_mutations_df = mutations_df_hg38[mutations_df_hg38[\"cancer_code\"].isin(cancer_types)]\n",
    "    print(\"total: \",cancer_mutations_df.shape)\n",
    "    \n",
    "    # deduplicate and mark those that are duplicates\n",
    "    cancer_mutations_df[\"duplicated_mutation\"] = cancer_mutations_df.duplicated([\"Chromosome\",\"hg38_start\",\"Tumor_Seq_Allele2\"], keep=False).astype(int)\n",
    "    cancer_mutations_df = cancer_mutations_df.drop_duplicates([\"Chromosome\",\"hg38_start\",\"Tumor_Seq_Allele2\"])\n",
    "    print(\"after deduplication: \",cancer_mutations_df.shape)\n",
    "    \n",
    "    # sort\n",
    "    cancer_mutations_df = cancer_mutations_df.sort_values([\"Chromosome\",\"hg38_start\"]).reset_index(drop=True)\n",
    "    \n",
    "    \n",
    "    # provide a mutation ID\n",
    "    cancer_mutations_df[\"mutation_id\"] = cancer_mutations_df.index\n",
    "    \n",
    "        \n",
    "    ## where to save\n",
    "    outfile = f\"cancer_specific_somatic_hg38/{cancer_name}_somatic_hg38.pkl\"\n",
    "    \n",
    "    print(cancer_mutations_df.head())\n",
    "    \n",
    "    # write into a file\n",
    "    write_pickle(cancer_mutations_df,outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7034dd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad6d7f11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### check 1\n",
    "# (mutations_df[\"Reference_Allele\"] == mutations_df[\"Tumor_Seq_Allele1\"]).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af7666ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### check 2\n",
    "# (mutations_df[\"i_NumCallers\"].astype(int) == mutations_df[\"i_Callers\"].apply(lambda lst:len(lst.split(\",\"))).astype(int)).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42606d3b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# ### check 3 - Start_position - 1 => Reference_Allele\n",
    "# f19 = Fasta(\"/illumina/scratch/deep_learning/akumar22/hg19_genome.fa\",sequence_always_upper=True)\n",
    "# mismatched_indices = []\n",
    "# for idx,row in mutations_df.iterrows():\n",
    "#     chm = str(row[\"Chromosome\"])\n",
    "#     start = int(row[\"Start_position\"])\n",
    "#     ref = str(row[\"Reference_Allele\"])\n",
    "#     if f19[chm][start-1] != ref:\n",
    "#         mismatched_indices.append(idx)\n",
    "#     if idx % 100000 == 1:\n",
    "#         print(idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2389be21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# mismatched_indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "096f9198",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### check 4 - strand\n",
    "# mutations_df[\"Strand\"].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
