{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1beb6d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a0ccf866",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/akumar22/anaconda3/envs/python38/lib/python3.10/site-packages/tqdm/auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from IPython.display import display, HTML\n",
    "import scipy.stats\n",
    "from pybedtools import BedTool\n",
    "import pickle\n",
    "from tqdm import tqdm\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "from seq2atac.stable import compute_gc_bed, one_hot_encode, read_pickle, compute_signal, write_pickle\n",
    "from seq2atac.analysis import get_promoterai_tss, get_cosmic_pancan_genes, fasta_file\n",
    "from seq2atac.analysis.enrichment_utils import create_pancancer_distribution_plots, create_pancancer_correlations, create_pancancer_valuecounts, create_pancancer_distribution_plots_discrete\n",
    "from seq2atac.analysis.sample_controls import matching_logic\n",
    "from seq2atac.analysis.mutation_utils import compute_vierstra_groups, compute_trinuc, ingene_indicator\n",
    "from seq2atac.analysis.mutation_processing_pipeline_utils import annotateNearestGene, annotateScore, filterannotatePeaks, annotatePhylopScore, annotatePhylopScorePeak, annotateBedfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2a7472ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "565"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pancan_genes = read_pickle(\"../process_tcga_icgc/ogtsg.pkl\")\n",
    "len(pancan_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "23a5c61c",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_tcga_cancers = [\"BLCA\",\"BRCA\",\"GBM\",\"COAD\",\"KIRC\",\"KIRP\",\"LUAD\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f554943",
   "metadata": {},
   "outputs": [],
   "source": [
    "ism_scores = pd.read_csv(\"../score_backups/ism_scores_genomewide/all_cancers_all_mutations.csv\",index_col=0)\n",
    "\n",
    "def merge_ism_scores(somatic_df,ism_df):\n",
    "    \n",
    "    sh_before = len(somatic_df)\n",
    "    somatic_df = somatic_df.merge(ism_df,how=\"left\")\n",
    "    assert sh_before == somatic_df.shape[0]\n",
    "    assert len(somatic_df[somatic_df[\"ref\"].isna()]) == 0\n",
    "    return somatic_df\n",
    "\n",
    "# ism_scores[\"sample\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0730e0c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BLCA': 0.30038172008,\n",
       " 'BRCA': 0.40271141012000006,\n",
       " 'GBM': 0.256060285,\n",
       " 'COAD': 0.2942905675,\n",
       " 'KIRC': 0.39631355571677007,\n",
       " 'KIRP': 0.2664290534,\n",
       " 'LUAD': 0.24252313195000003,\n",
       " 'RECA': 0.39631355571677007}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak_proba_thresh = read_pickle(\"../process_tcga_icgc/peak_proba_thresh.pkl\")\n",
    "peak_proba_thresh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "30ee187e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BLCA\n",
      "(155700, 31)\n",
      "(155700, 49)\n",
      "(155700, 53)\n",
      "(155700, 54)\n",
      "(18438, 55)\n",
      "1129\n",
      "(11393, 55)\n",
      "BRCA\n",
      "(376478, 31)\n",
      "(376478, 49)\n",
      "(376478, 53)\n",
      "(376478, 54)\n",
      "(55645, 55)\n",
      "3528\n",
      "(33464, 55)\n",
      "GBM\n",
      "(89159, 31)\n",
      "(89159, 49)\n",
      "(89159, 53)\n",
      "(89159, 54)\n",
      "(8370, 55)\n",
      "486\n",
      "(4989, 55)\n",
      "COAD\n",
      "(1165711, 31)\n",
      "(1165711, 49)\n",
      "(1165711, 53)\n",
      "(1165711, 54)\n",
      "(84645, 55)\n",
      "4959\n",
      "(43990, 55)\n",
      "KIRC\n",
      "(223811, 31)\n",
      "(223811, 49)\n",
      "(223811, 53)\n",
      "(223811, 54)\n",
      "(12289, 55)\n",
      "729\n",
      "(5492, 55)\n",
      "KIRP\n",
      "(56691, 31)\n",
      "(56691, 49)\n",
      "(56691, 53)\n",
      "(56691, 54)\n",
      "(5587, 55)\n",
      "332\n",
      "(3183, 55)\n",
      "LUAD\n",
      "(377986, 31)\n",
      "(377986, 49)\n",
      "(377986, 53)\n",
      "(377986, 54)\n",
      "(30787, 55)\n",
      "1807\n",
      "(15102, 55)\n"
     ]
    }
   ],
   "source": [
    "somatic_df_dict = {}\n",
    "for cancer_name in all_tcga_cancers:\n",
    "    print(cancer_name)\n",
    "\n",
    "    ### read the processed somatic dataframes (this ensures within peaks)\n",
    "    somatic_df = read_pickle(f\"../somatic_filtered/{cancer_name}_filtered_annotated_somatic.pkl\")\n",
    "    assert len(somatic_df.drop_duplicates([\"Chromosome\",\"hg38_start\",\"Tumor_Seq_Allele2\"])) == len(somatic_df)\n",
    "    print(somatic_df.shape)    \n",
    "    \n",
    "    ### annotate model scores\n",
    "\n",
    "    somatic_df = annotateScore(somatic_df,\n",
    "                               f\"../score_backups/mutation_centered_genomewide/{cancer_name}/fold_*_somatic_mutation_centered.pkl\",\n",
    "                              remove_foldwise=False)\n",
    "    somatic_df = somatic_df.rename(columns={\"proba_ref\":\"proba_ref_mutation_centered\",\n",
    "                                            \"proba_alt\":\"proba_alt_mutation_centered\"})\n",
    "    somatic_df[\"diff_mutation_centered\"] = somatic_df[\"proba_ref_mutation_centered\"] - somatic_df[\"proba_alt_mutation_centered\"]\n",
    "    for fold in range(5):\n",
    "        somatic_df[f\"fold_{fold}_diff_mutation_centered\"] = somatic_df[f\"proba_ref_{fold}\"] - somatic_df[f\"proba_alt_{fold}\"]\n",
    "    print(somatic_df.shape)\n",
    "    \n",
    "    ### annotate ism scores\n",
    "    somatic_df = merge_ism_scores(somatic_df,\n",
    "                                  ism_scores[ism_scores[\"sample\"]==cancer_name.lower()])\n",
    "    print(somatic_df.shape)\n",
    "    \n",
    "    \n",
    "    ### annotate pancan\n",
    "    somatic_df[\"closest_ogtsg\"] = somatic_df[\"gene\"].apply(lambda x : ingene_indicator(x, pancan_genes))\n",
    "    print(somatic_df.shape)\n",
    "    \n",
    "    ### outside peaks, but < 1000\n",
    "    somatic_df = annotateBedfile(somatic_df, \n",
    "                                 f\"/illumina/scratch/deep_learning/akumar22/TCGA/mutations_scoring/master_files/cancer_peaks_500/{cancer_name}_peaks_indexed.bed\", \n",
    "                                 \"inside_peaks\")\n",
    "    somatic_df = somatic_df[(somatic_df[\"inside_peaks\"]==0) & (somatic_df[\"distance_to_summit\"] <= 1000)]\n",
    "    print(somatic_df.shape)\n",
    "    print(len(somatic_df[somatic_df[\"closest_ogtsg\"]==1]))\n",
    "    \n",
    "    ### \n",
    "    minpeak_proba = peak_proba_thresh[cancer_name]\n",
    "    somatic_df = somatic_df[somatic_df[\"proba_ref_mutation_centered\"] > minpeak_proba]\n",
    "    print(somatic_df.shape)\n",
    "    \n",
    "    \n",
    "    ### some annotations for matching\n",
    "    somatic_df[\"gc_2m\"] = somatic_df[\"gc_mutation\"].apply(lambda x:round(x,2))\n",
    "    somatic_df[\"gc_1m\"] = somatic_df[\"gc_mutation\"].apply(lambda x:round(x,1))\n",
    "    somatic_df[\"distance_to_summit_discrete\"] = somatic_df[\"distance_to_summit\"].apply(lambda x:round(x,-1))\n",
    "    somatic_df[\"distance_to_tss_discrete\"] = somatic_df[\"distance_to_tss\"].apply(lambda x:round(x,-4))\n",
    "    \n",
    "    assert len(somatic_df.drop_duplicates([\"Chromosome\",\"hg38_start\",\"Tumor_Seq_Allele2\"])) == len(somatic_df)    \n",
    "    somatic_df_dict[cancer_name] = somatic_df\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fe8fa250",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Store these dataframes\n",
    "write_pickle(somatic_df_dict,\"somatic_df_dict_gof.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95bf918f",
   "metadata": {},
   "source": [
    "### Generate Matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ea2fd591",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_match(somatic_df_dict,genelist):\n",
    "    \n",
    "    all_tcga_cancers = list(somatic_df_dict.keys())\n",
    "    \n",
    "    ### gene annotation\n",
    "    for cancer_name in all_tcga_cancers:\n",
    "        somatic_df = somatic_df_dict[cancer_name].copy()\n",
    "        print(somatic_df.shape)\n",
    "\n",
    "        somatic_df[\"closest_pancan\"] = somatic_df[\"gene\"].apply(lambda x : ingene_indicator(x, genelist))\n",
    "\n",
    "        print(somatic_df.shape)\n",
    "        somatic_df_dict[cancer_name] = somatic_df   \n",
    "        \n",
    "    ### matching\n",
    "    pancan_matched_dict = {}\n",
    "    nonpancan_matched_dict = {}\n",
    "\n",
    "    for cancer_name in all_tcga_cancers:\n",
    "        print(cancer_name)\n",
    "        somatic_df = somatic_df_dict[cancer_name].copy()\n",
    "        df1, df2 = matching_logic(somatic_df[somatic_df[\"closest_pancan\"]==1],\n",
    "                                             somatic_df[(somatic_df[\"closest_pancan\"]==0) & (somatic_df[\"closest_ogtsg\"]==0)],\n",
    "                                             levels=[[\"trinuc\",\"gc_2m\",\"distance_to_summit_discrete\",\"distance_to_tss_discrete\"],\n",
    "                                                     [\"gc_2m\",\"distance_to_summit_discrete\",\"distance_to_tss_discrete\"],\n",
    "                                                     [\"gc_2m\",\"distance_to_summit_discrete\"],\n",
    "                                                     [\"gc_2m\"],\n",
    "                                                     [\"gc_1m\"]],\n",
    "                                             verbose=False)\n",
    "\n",
    "        pancan_matched_dict[cancer_name] = df1\n",
    "        nonpancan_matched_dict[cancer_name] = df2\n",
    "        \n",
    "    ### sanity\n",
    "    df_stats = pd.DataFrame()\n",
    "    for cancer_name in all_tcga_cancers:\n",
    "        print(cancer_name)\n",
    "        somatic_df = somatic_df_dict[cancer_name].copy()\n",
    "        df_stats.loc[cancer_name,\"total somatic\"] = len(somatic_df)\n",
    "        df_stats.loc[cancer_name,\"pancan\"] = len(somatic_df[(somatic_df[\"closest_pancan\"]==1)])\n",
    "        df_stats.loc[cancer_name,\"nonpancan\"] = len(somatic_df[(somatic_df[\"closest_pancan\"]==0)])\n",
    "        df_stats.loc[cancer_name,\"matched pancan\"] = len(pancan_matched_dict[cancer_name])\n",
    "        df_stats.loc[cancer_name,\"matched not near pancan\"] = len(nonpancan_matched_dict[cancer_name])\n",
    "        \n",
    "#         assert len(somatic_df[(somatic_df[\"closest_pancan\"]==1)]) == len(pancan_matched_dict[cancer_name])\n",
    "#         assert len(pancan_matched_dict[cancer_name]) == len(nonpancan_matched_dict[cancer_name])\n",
    "        \n",
    "    display(df_stats.astype(int))\n",
    " \n",
    "    \n",
    "    return pancan_matched_dict, nonpancan_matched_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c2d993f5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "565\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>694</td>\n",
       "      <td>10699</td>\n",
       "      <td>694</td>\n",
       "      <td>694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2087</td>\n",
       "      <td>31377</td>\n",
       "      <td>2087</td>\n",
       "      <td>2087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>296</td>\n",
       "      <td>4693</td>\n",
       "      <td>296</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2584</td>\n",
       "      <td>41406</td>\n",
       "      <td>2584</td>\n",
       "      <td>2584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>357</td>\n",
       "      <td>5135</td>\n",
       "      <td>357</td>\n",
       "      <td>357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>198</td>\n",
       "      <td>2985</td>\n",
       "      <td>198</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>896</td>\n",
       "      <td>14206</td>\n",
       "      <td>896</td>\n",
       "      <td>896</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     694      10699             694   \n",
       "BRCA          33464    2087      31377            2087   \n",
       "GBM            4989     296       4693             296   \n",
       "COAD          43990    2584      41406            2584   \n",
       "KIRC           5492     357       5135             357   \n",
       "KIRP           3183     198       2985             198   \n",
       "LUAD          15102     896      14206             896   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      694  \n",
       "BRCA                     2087  \n",
       "GBM                       296  \n",
       "COAD                     2584  \n",
       "KIRC                      357  \n",
       "KIRP                      198  \n",
       "LUAD                      896  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(len(pancan_genes))\n",
    "pancan_matched_dict, nonpancan_matched_dict = get_match(somatic_df_dict,pancan_genes)\n",
    "write_pickle(pancan_matched_dict,\"pancan_matched_dict_ogtsg.pkl\")\n",
    "write_pickle(nonpancan_matched_dict,\"nonpancan_matched_dict_ogtsg.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f7fab4",
   "metadata": {},
   "source": [
    "### 5 random pli matched genesets - also total mutations matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e87dada4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>717</td>\n",
       "      <td>10676</td>\n",
       "      <td>717</td>\n",
       "      <td>717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2321</td>\n",
       "      <td>31143</td>\n",
       "      <td>2321</td>\n",
       "      <td>2321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>353</td>\n",
       "      <td>4636</td>\n",
       "      <td>353</td>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2923</td>\n",
       "      <td>41067</td>\n",
       "      <td>2923</td>\n",
       "      <td>2923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>387</td>\n",
       "      <td>5105</td>\n",
       "      <td>387</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>209</td>\n",
       "      <td>2974</td>\n",
       "      <td>209</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>1025</td>\n",
       "      <td>14077</td>\n",
       "      <td>1025</td>\n",
       "      <td>1025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     717      10676             717   \n",
       "BRCA          33464    2321      31143            2321   \n",
       "GBM            4989     353       4636             353   \n",
       "COAD          43990    2923      41067            2923   \n",
       "KIRC           5492     387       5105             387   \n",
       "KIRP           3183     209       2974             209   \n",
       "LUAD          15102    1025      14077            1025   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      717  \n",
       "BRCA                     2321  \n",
       "GBM                       353  \n",
       "COAD                     2923  \n",
       "KIRC                      387  \n",
       "KIRP                      209  \n",
       "LUAD                     1025  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>703</td>\n",
       "      <td>10690</td>\n",
       "      <td>703</td>\n",
       "      <td>703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2140</td>\n",
       "      <td>31324</td>\n",
       "      <td>2140</td>\n",
       "      <td>2140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>305</td>\n",
       "      <td>4684</td>\n",
       "      <td>305</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2690</td>\n",
       "      <td>41300</td>\n",
       "      <td>2690</td>\n",
       "      <td>2690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>387</td>\n",
       "      <td>5105</td>\n",
       "      <td>387</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>223</td>\n",
       "      <td>2960</td>\n",
       "      <td>223</td>\n",
       "      <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>906</td>\n",
       "      <td>14196</td>\n",
       "      <td>906</td>\n",
       "      <td>906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     703      10690             703   \n",
       "BRCA          33464    2140      31324            2140   \n",
       "GBM            4989     305       4684             305   \n",
       "COAD          43990    2690      41300            2690   \n",
       "KIRC           5492     387       5105             387   \n",
       "KIRP           3183     223       2960             223   \n",
       "LUAD          15102     906      14196             906   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      703  \n",
       "BRCA                     2140  \n",
       "GBM                       305  \n",
       "COAD                     2690  \n",
       "KIRC                      387  \n",
       "KIRP                      223  \n",
       "LUAD                      906  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>843</td>\n",
       "      <td>10550</td>\n",
       "      <td>843</td>\n",
       "      <td>843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2458</td>\n",
       "      <td>31006</td>\n",
       "      <td>2458</td>\n",
       "      <td>2458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>430</td>\n",
       "      <td>4559</td>\n",
       "      <td>430</td>\n",
       "      <td>430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>3172</td>\n",
       "      <td>40818</td>\n",
       "      <td>3172</td>\n",
       "      <td>3172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>364</td>\n",
       "      <td>5128</td>\n",
       "      <td>364</td>\n",
       "      <td>364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>219</td>\n",
       "      <td>2964</td>\n",
       "      <td>219</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>1101</td>\n",
       "      <td>14001</td>\n",
       "      <td>1101</td>\n",
       "      <td>1101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     843      10550             843   \n",
       "BRCA          33464    2458      31006            2458   \n",
       "GBM            4989     430       4559             430   \n",
       "COAD          43990    3172      40818            3172   \n",
       "KIRC           5492     364       5128             364   \n",
       "KIRP           3183     219       2964             219   \n",
       "LUAD          15102    1101      14001            1101   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      843  \n",
       "BRCA                     2458  \n",
       "GBM                       430  \n",
       "COAD                     3172  \n",
       "KIRC                      364  \n",
       "KIRP                      219  \n",
       "LUAD                     1101  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>776</td>\n",
       "      <td>10617</td>\n",
       "      <td>776</td>\n",
       "      <td>776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2404</td>\n",
       "      <td>31060</td>\n",
       "      <td>2404</td>\n",
       "      <td>2404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>375</td>\n",
       "      <td>4614</td>\n",
       "      <td>375</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>3047</td>\n",
       "      <td>40943</td>\n",
       "      <td>3047</td>\n",
       "      <td>3047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>365</td>\n",
       "      <td>5127</td>\n",
       "      <td>365</td>\n",
       "      <td>365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>226</td>\n",
       "      <td>2957</td>\n",
       "      <td>226</td>\n",
       "      <td>226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>1071</td>\n",
       "      <td>14031</td>\n",
       "      <td>1071</td>\n",
       "      <td>1071</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     776      10617             776   \n",
       "BRCA          33464    2404      31060            2404   \n",
       "GBM            4989     375       4614             375   \n",
       "COAD          43990    3047      40943            3047   \n",
       "KIRC           5492     365       5127             365   \n",
       "KIRP           3183     226       2957             226   \n",
       "LUAD          15102    1071      14031            1071   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      776  \n",
       "BRCA                     2404  \n",
       "GBM                       375  \n",
       "COAD                     3047  \n",
       "KIRC                      365  \n",
       "KIRP                      226  \n",
       "LUAD                     1071  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1114\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>701</td>\n",
       "      <td>10692</td>\n",
       "      <td>701</td>\n",
       "      <td>701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2182</td>\n",
       "      <td>31282</td>\n",
       "      <td>2182</td>\n",
       "      <td>2182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>338</td>\n",
       "      <td>4651</td>\n",
       "      <td>338</td>\n",
       "      <td>338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2830</td>\n",
       "      <td>41160</td>\n",
       "      <td>2830</td>\n",
       "      <td>2830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>409</td>\n",
       "      <td>5083</td>\n",
       "      <td>409</td>\n",
       "      <td>409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>217</td>\n",
       "      <td>2966</td>\n",
       "      <td>217</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>983</td>\n",
       "      <td>14119</td>\n",
       "      <td>983</td>\n",
       "      <td>983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     701      10692             701   \n",
       "BRCA          33464    2182      31282            2182   \n",
       "GBM            4989     338       4651             338   \n",
       "COAD          43990    2830      41160            2830   \n",
       "KIRC           5492     409       5083             409   \n",
       "KIRP           3183     217       2966             217   \n",
       "LUAD          15102     983      14119             983   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      701  \n",
       "BRCA                     2182  \n",
       "GBM                       338  \n",
       "COAD                     2830  \n",
       "KIRC                      409  \n",
       "KIRP                      217  \n",
       "LUAD                      983  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    pli_matched_genes = read_pickle(f\"../process_tcga_icgc/pli_matched_genes_x2{i}.pkl\")\n",
    "    print(len(pli_matched_genes))\n",
    "    \n",
    "    assert len(set(pli_matched_genes).intersection(set(pancan_genes))) == 0\n",
    "    \n",
    "    pancan_matched_dict_pli, nonpancan_matched_dict_pli = get_match(somatic_df_dict,pli_matched_genes)\n",
    "    \n",
    "    write_pickle(pancan_matched_dict_pli,f\"pancan_matched_dict_pli_x2{i}.pkl\")\n",
    "    write_pickle(nonpancan_matched_dict_pli,f\"nonpancan_matched_dict_pli_x2{i}.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ab57f347",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for i in range(5):\n",
    "#     pli_matched_genes = read_pickle(f\"../process_tcga_icgc/pli_matched_oncogenes_x2{i}.pkl\")\n",
    "#     print(len(pli_matched_genes))\n",
    "    \n",
    "#     assert len(set(pli_matched_genes).intersection(set(pancan_genes))) == 0\n",
    "    \n",
    "#     pancan_matched_dict_pli, nonpancan_matched_dict_pli = get_match(somatic_df_dict,pli_matched_genes)\n",
    "    \n",
    "#     write_pickle(pancan_matched_dict_pli,f\"pancan_matched_dict_pli_onco_x2{i}.pkl\")\n",
    "#     write_pickle(nonpancan_matched_dict_pli,f\"nonpancan_matched_dict_pli_onco_x2{i}.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "839ea9ff",
   "metadata": {},
   "source": [
    "### 5 random genesets - also total mutations matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2c6de6c4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>749</td>\n",
       "      <td>10644</td>\n",
       "      <td>749</td>\n",
       "      <td>749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>1965</td>\n",
       "      <td>31499</td>\n",
       "      <td>1965</td>\n",
       "      <td>1965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>283</td>\n",
       "      <td>4706</td>\n",
       "      <td>283</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2532</td>\n",
       "      <td>41458</td>\n",
       "      <td>2531</td>\n",
       "      <td>2531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>328</td>\n",
       "      <td>5164</td>\n",
       "      <td>328</td>\n",
       "      <td>328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>192</td>\n",
       "      <td>2991</td>\n",
       "      <td>192</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>903</td>\n",
       "      <td>14199</td>\n",
       "      <td>903</td>\n",
       "      <td>903</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     749      10644             749   \n",
       "BRCA          33464    1965      31499            1965   \n",
       "GBM            4989     283       4706             283   \n",
       "COAD          43990    2532      41458            2531   \n",
       "KIRC           5492     328       5164             328   \n",
       "KIRP           3183     192       2991             192   \n",
       "LUAD          15102     903      14199             903   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      749  \n",
       "BRCA                     1965  \n",
       "GBM                       283  \n",
       "COAD                     2531  \n",
       "KIRC                      328  \n",
       "KIRP                      192  \n",
       "LUAD                      903  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>667</td>\n",
       "      <td>10726</td>\n",
       "      <td>667</td>\n",
       "      <td>667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2033</td>\n",
       "      <td>31431</td>\n",
       "      <td>2033</td>\n",
       "      <td>2033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>309</td>\n",
       "      <td>4680</td>\n",
       "      <td>309</td>\n",
       "      <td>309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2687</td>\n",
       "      <td>41303</td>\n",
       "      <td>2687</td>\n",
       "      <td>2687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>340</td>\n",
       "      <td>5152</td>\n",
       "      <td>340</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>165</td>\n",
       "      <td>3018</td>\n",
       "      <td>165</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>929</td>\n",
       "      <td>14173</td>\n",
       "      <td>929</td>\n",
       "      <td>929</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     667      10726             667   \n",
       "BRCA          33464    2033      31431            2033   \n",
       "GBM            4989     309       4680             309   \n",
       "COAD          43990    2687      41303            2687   \n",
       "KIRC           5492     340       5152             340   \n",
       "KIRP           3183     165       3018             165   \n",
       "LUAD          15102     929      14173             929   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      667  \n",
       "BRCA                     2033  \n",
       "GBM                       309  \n",
       "COAD                     2687  \n",
       "KIRC                      340  \n",
       "KIRP                      165  \n",
       "LUAD                      929  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>654</td>\n",
       "      <td>10739</td>\n",
       "      <td>654</td>\n",
       "      <td>654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>1882</td>\n",
       "      <td>31582</td>\n",
       "      <td>1882</td>\n",
       "      <td>1882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>275</td>\n",
       "      <td>4714</td>\n",
       "      <td>275</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2576</td>\n",
       "      <td>41414</td>\n",
       "      <td>2576</td>\n",
       "      <td>2576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>313</td>\n",
       "      <td>5179</td>\n",
       "      <td>313</td>\n",
       "      <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>201</td>\n",
       "      <td>2982</td>\n",
       "      <td>201</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>843</td>\n",
       "      <td>14259</td>\n",
       "      <td>843</td>\n",
       "      <td>843</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     654      10739             654   \n",
       "BRCA          33464    1882      31582            1882   \n",
       "GBM            4989     275       4714             275   \n",
       "COAD          43990    2576      41414            2576   \n",
       "KIRC           5492     313       5179             313   \n",
       "KIRP           3183     201       2982             201   \n",
       "LUAD          15102     843      14259             843   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      654  \n",
       "BRCA                     1882  \n",
       "GBM                       275  \n",
       "COAD                     2576  \n",
       "KIRC                      313  \n",
       "KIRP                      201  \n",
       "LUAD                      843  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>659</td>\n",
       "      <td>10734</td>\n",
       "      <td>659</td>\n",
       "      <td>659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2062</td>\n",
       "      <td>31402</td>\n",
       "      <td>2062</td>\n",
       "      <td>2062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>362</td>\n",
       "      <td>4627</td>\n",
       "      <td>362</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2798</td>\n",
       "      <td>41192</td>\n",
       "      <td>2798</td>\n",
       "      <td>2798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>319</td>\n",
       "      <td>5173</td>\n",
       "      <td>319</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>178</td>\n",
       "      <td>3005</td>\n",
       "      <td>178</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>922</td>\n",
       "      <td>14180</td>\n",
       "      <td>922</td>\n",
       "      <td>922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     659      10734             659   \n",
       "BRCA          33464    2062      31402            2062   \n",
       "GBM            4989     362       4627             362   \n",
       "COAD          43990    2798      41192            2798   \n",
       "KIRC           5492     319       5173             319   \n",
       "KIRP           3183     178       3005             178   \n",
       "LUAD          15102     922      14180             922   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      659  \n",
       "BRCA                     2062  \n",
       "GBM                       362  \n",
       "COAD                     2798  \n",
       "KIRC                      319  \n",
       "KIRP                      178  \n",
       "LUAD                      922  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1130\n",
      "(11393, 59)\n",
      "(11393, 59)\n",
      "(33464, 59)\n",
      "(33464, 59)\n",
      "(4989, 59)\n",
      "(4989, 59)\n",
      "(43990, 59)\n",
      "(43990, 59)\n",
      "(5492, 59)\n",
      "(5492, 59)\n",
      "(3183, 59)\n",
      "(3183, 59)\n",
      "(15102, 59)\n",
      "(15102, 59)\n",
      "BLCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "BRCA\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "GBM\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "COAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRC\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "KIRP\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "LUAD\n",
      "Grouping by:  ['trinuc', 'gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete', 'distance_to_tss_discrete']\n",
      "Grouping by:  ['gc_2m', 'distance_to_summit_discrete']\n",
      "Grouping by:  ['gc_2m']\n",
      "Grouping by:  ['gc_1m']\n",
      "BLCA\n",
      "BRCA\n",
      "GBM\n",
      "COAD\n",
      "KIRC\n",
      "KIRP\n",
      "LUAD\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>total somatic</th>\n",
       "      <th>pancan</th>\n",
       "      <th>nonpancan</th>\n",
       "      <th>matched pancan</th>\n",
       "      <th>matched not near pancan</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BLCA</th>\n",
       "      <td>11393</td>\n",
       "      <td>638</td>\n",
       "      <td>10755</td>\n",
       "      <td>638</td>\n",
       "      <td>638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA</th>\n",
       "      <td>33464</td>\n",
       "      <td>2026</td>\n",
       "      <td>31438</td>\n",
       "      <td>2026</td>\n",
       "      <td>2026</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GBM</th>\n",
       "      <td>4989</td>\n",
       "      <td>278</td>\n",
       "      <td>4711</td>\n",
       "      <td>278</td>\n",
       "      <td>278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COAD</th>\n",
       "      <td>43990</td>\n",
       "      <td>2871</td>\n",
       "      <td>41119</td>\n",
       "      <td>2871</td>\n",
       "      <td>2871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRC</th>\n",
       "      <td>5492</td>\n",
       "      <td>314</td>\n",
       "      <td>5178</td>\n",
       "      <td>314</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRP</th>\n",
       "      <td>3183</td>\n",
       "      <td>182</td>\n",
       "      <td>3001</td>\n",
       "      <td>182</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LUAD</th>\n",
       "      <td>15102</td>\n",
       "      <td>991</td>\n",
       "      <td>14111</td>\n",
       "      <td>990</td>\n",
       "      <td>990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      total somatic  pancan  nonpancan  matched pancan  \\\n",
       "BLCA          11393     638      10755             638   \n",
       "BRCA          33464    2026      31438            2026   \n",
       "GBM            4989     278       4711             278   \n",
       "COAD          43990    2871      41119            2871   \n",
       "KIRC           5492     314       5178             314   \n",
       "KIRP           3183     182       3001             182   \n",
       "LUAD          15102     991      14111             990   \n",
       "\n",
       "      matched not near pancan  \n",
       "BLCA                      638  \n",
       "BRCA                     2026  \n",
       "GBM                       278  \n",
       "COAD                     2871  \n",
       "KIRC                      314  \n",
       "KIRP                      182  \n",
       "LUAD                      990  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    random_genes = read_pickle(f\"../process_tcga_icgc/random_genes_{i}.pkl\")\n",
    "    print(len(random_genes))\n",
    "    \n",
    "    assert len(set(random_genes).intersection(set(pancan_genes))) == 0\n",
    "    \n",
    "    pancan_matched_dict_random, nonpancan_matched_dict_random = get_match(somatic_df_dict,random_genes)\n",
    "    \n",
    "    write_pickle(pancan_matched_dict_random,f\"pancan_matched_dict_random_{i}.pkl\")\n",
    "    write_pickle(nonpancan_matched_dict_random,f\"nonpancan_matched_dict_random_{i}.pkl\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e536e991",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done\n"
     ]
    }
   ],
   "source": [
    "print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b2baf48",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
